File: src/schemas/metadata/MetadataProperty.ts
import { ClassProperties } from "../../typings/ClassProperties";

import { IProtobufProperty } from "../protobuf/IProtobufProperty";
import { IJsDocTagInfo } from "./IJsDocTagInfo";
import { IMetadataDictionary } from "./IMetadataDictionary";
import { IMetadataProperty } from "./IMetadataProperty";
import { Metadata } from "./Metadata";

export class MetadataProperty {
  public readonly key: Metadata;
  public readonly value: Metadata;
  public readonly description: string | null;
  public readonly jsDocTags: IJsDocTagInfo[];

  public of_protobuf_?: IProtobufProperty;

  /* -----------------------------------------------------------
        CONSTRUCTORS
    ----------------------------------------------------------- */
  /**
   * @hidden
   */
  private constructor(props: ClassProperties<MetadataProperty>) {
    this.key = props.key;
    this.value = props.value;
    this.description = props.description;
    this.jsDocTags = props.jsDocTags;
  }

  /**
   * @internal
   */
  public static create(
    props: ClassProperties<MetadataProperty>,
  ): MetadataProperty {
    return new MetadataProperty(props);
  }

  /**
   * @internal
   */
  public static from(property: IMetadataProperty, dict: IMetadataDictionary) {
    return MetadataProperty.create({
      key: Metadata.from(property.key, dict),
      value: Metadata.from(property.value, dict),
      description: property.description,
      jsDocTags: property.jsDocTags.slice(),
    });
  }

  public toJSON(): IMetadataProperty {
    return {
      key: this.key.toJSON(),
      value: this.value.toJSON(),
      description: this.description,
      jsDocTags: this.jsDocTags,
    };
  }
}

File: packages/typia/native/core/schemas/metadata/MetadataProperty.go
package metadata

import nativeprotobuf "github.com/samchon/typia/packages/typia/native/core/schemas/protobuf"

type IMetadataSchema_IProperty struct {
  Key         *IMetadataSchema
  Value       *IMetadataSchema
  Description *string
  JsDocTags   []IJsDocTagInfo
  Mutability  *string
}

type MetadataProperty struct {
  Key          *MetadataSchema
  Value        *MetadataSchema
  Description  *string
  JsDocTags    []IJsDocTagInfo
  Mutability   *string
  Of_protobuf_ *nativeprotobuf.IProtobufProperty
}

func MetadataProperty_create(props MetadataProperty) *MetadataProperty {
  return &MetadataProperty{
    Key:         props.Key,
    Value:       props.Value,
    Description: props.Description,
    JsDocTags:   append([]IJsDocTagInfo{}, props.JsDocTags...),
    Mutability:  props.Mutability,
  }
}

func MetadataProperty_from(property IMetadataSchema_IProperty, dict IMetadataDictionary) *MetadataProperty {
  return MetadataProperty_create(MetadataProperty{
    Key:         MetadataSchema_from(property.Key, dict),
    Value:       MetadataSchema_from(property.Value, dict),
    Description: property.Description,
    JsDocTags:   append([]IJsDocTagInfo{}, property.JsDocTags...),
    Mutability:  property.Mutability,
  })
}

func (obj *MetadataProperty) ToJSON() IMetadataSchema_IProperty {
  return IMetadataSchema_IProperty{
    Key:         obj.Key.ToJSON(),
    Value:       obj.Value.ToJSON(),
    Description: obj.Description,
    JsDocTags:   append([]IJsDocTagInfo{}, obj.JsDocTags...),
    Mutability:  obj.Mutability,
  }
}
